### PR TITLE
Add Makefile and Boot Image Format template

### PR DESCRIPTION
**Pull request by lool**
Hi,

This allows building and installing all necessary files; note that I picked a /lib/firmware/xilinx/xilinx-app-$board-$name hierarchy.

Best,
LM

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+INSTALLDIR := $(DESTDIR)/lib/firmware/xilinx
+
+install: $(BINS) $(DTBOS) $(JSONS)
+	for f in $^; do \
+	    file=$$(basename $$f); \
+	    app=$$(basename $$(dirname $$f)); \
+	    board=$$(basename $$(dirname $$(dirname $$f))); \
+	    install -D -m 644 \
+	        $$f $(INSTALLDIR)/xilinx-app-$$board-$$app/$$file; \
```
Comment (chkohn) [inline Makefile]: to align with yocto naming convention, I suggest to use 
`$(INSTALLDIR)/$$board-$$app/$$file`
instead.

**Review by chkohn — COMMENTED** [review]
@lool thank you for submitting this patch.

this works great if a user wants to install all fw artifacts from this repo. I'm not sure how you are intending to package it but i assume it would be one package for all fw artifacts based on how this makefile is written. in yocto we are creating a separate fw package for each app e.g. kv260-smartcam. Then the corresponding app packagegroup will only pull in the app specific fw package rather than the superset. as apps have different release cycles and some of the fw artifacts in the repo may have not been updated/verified yet, I think it will be better to package and release them individually once ready. we just didn't want to create a separate repo for each app-based fw.

also, one inline comment on existing install dir naming convention.

**Comment by chkohn** [comment]
@lool one more thing and this can wait till the kv260-apps release.

Looking at https://github.com/Xilinx/kria-apps-firmware/tree/xlnx_rel_v2022.1/kv260/smartcam as an example, there are two more sets of artifacts that need to be installed:
1. xclbin file - this gets installed into the same directory as the bit.bin, dtbo and json file (if present)
2. Vitis AI models - these are ML model binaries that get installed into /opt/xilinx/\<app-name\>/share/vitis_ai_library/models (if present)

One or both can apply to each of the featured apps depending on whether they are using Vitis acceleration flow (xclbin) and/or Vitis AI (ML models). Most kv260 apps use one of them or both. Even the kr260 mv-camera app will have an xclbin file in a future iteration.

**Comment by vmayoral** [comment]
> Looking at https://github.com/Xilinx/kria-apps-firmware/tree/xlnx_rel_v2022.1/kv260/smartcam as an example, there are two more sets of artifacts that need to be installed:

I can confirm this is the case. See https://github.com/ros-acceleration/ament_vitis/blob/main/cmake/vitis_acceleration_kernel.cmake#L322-L372 for a CMake example of the `dfx-mgr` artifacts required in previous Vitis release versions.

**Comment by lool** [comment]
@chkohn Thanks for the review; on your various points:

1. one or multiple packages: actually, it's a single source package but in the packaging I'm splitting the files into multiple binary packages which is consistent with this being a single upstream module on github and the fact that each demo app needs this or that firmware to work so I think we're good

1. install path: makes sense for this Makefile to default to whatever you prefer targeting for e.g. a petalinux or yocto context; I can override this in the packaging to avoid separate namespaces for sideloaded apps proposed by Xilinx as debs vs local assets vs other means of delivering the firmware; I can change this to `/lib/firmware/xilinx/${board}-${appname}`

1. xclbin: would you mind confirming the exact preferred invocation and could you confirm if this is urgent or not? is there a minimum XRT version to use?

1. Vitis AI models: this is the part I'm least comfortable with as we are packaging Vitis AI libs for Ubuntu, but I don't fully trust it's split and configured correctly yet – and I don't want to ship OS files under /opt; which pieces of software would expect to find files there? programs linked to Vitis AI libs? is this a search path that can easily be overriden? I think this Makefile should default to whatever is convenient for Yocto / PetaLinux but perhaps an OS level default could usr/lib/triplet/vitis-ai-models and we'd make sure Vitis-AI searches there. Or usr/share/vitis-ai-models if these are arch-independent (you can use the exact same pre-built model file whether your board is running arm64 or armhf versions of Vitis AI, or amd64 if you're on a server talking to a PCI express attached board or something)

**Comment by chkohn** [comment]
@lool 
1. I say this when looking at the package details on launchpad this morning. this looks good, just wanted to make sure the binary packages are broken out per app. thanks for confirming.
2. the yocto flow is not using the makefile, we have a base class implemented in yocto that know how to deal with firmware packages. so you can keep it as so it matches the default Ubuntu installation paths. My comment was more wrt matching the yocto paths but we discussed and agreed this morning. so this is good too.
3. for xclbin, there is not processing needed, just a simple copy like the json file. I'm not sure on the coupling of XRT and xclbin versions. we typically update them together and in sync with the vivado/vitis tools version. this is first needed for the kr260-perception app I believe which will follow soon after the initial kr260-apps. and it is also needed for the kv260-apps. so not super urgent.

will comment on 4 separately (this is also not urgent for initial kr260 apps).

**Comment by lool** [comment]
I've addressed 2 and 3 with above commits

**Review by chkohn — APPROVED** [review]
thanks @lool

**Comment by chkohn** [comment]
> Vitis AI models: this is the part I'm least comfortable with as we are packaging Vitis AI libs for Ubuntu, but I don't fully trust it's split and configured correctly yet – and I don't want to ship OS files under /opt; which pieces of software would expect to find files there? programs linked to Vitis AI libs? is this a search path that can easily be overriden? I think this Makefile should default to whatever is convenient for Yocto / PetaLinux but perhaps an OS level default could usr/lib/triplet/vitis-ai-models and we'd make sure Vitis-AI searches there. Or usr/share/vitis-ai-models if these are arch-independent (you can use the exact same pre-built model file whether your board is running arm64 or armhf versions of Vitis AI, or amd64 if you're on a server talking to a PCI express attached board or something)

@lool yes, applications linked to Vitis AI libs typically require AI model binaries for running AI workloads. I'm not entirely sure but I think there are several ways to identify the search path: conf file, env variable, maybe also programmatically? Then there are also different sw layers of interacting with vitis AI: one is using the VAI runtime which is more low-level, another is using VVAS which provides gstreamer plugins on top of VAI (amongst other things). In VVAS there is a per-plugin json file which I believe sets the model path for that plugin.

We are installing the models into /opt because they are tied to the application and firmware e.g. the bitstream includes a certain configuration of the DPU which is the AI processing IP and the models are compiled based on that configuration and may not work on other configurations. Therefore we don't want to install them into the OS as there are many permutations of models and configurations possible (yay FPGA flexibility :)). 

Even for petalinux/yocto, we are currently installing the apps and the models into /opt/xilinx/<app> separating it from the OS. Only the firmware is installed into /lib/firmware/xilinx/<app>.

**Comment by lool** [comment]
If this is application-specific, perhaps these should live with the apps? But I get the impression it's also tied to the bitstream(s) you're currently running and that you might want to update the bitstreams along the corresponding models along the corresponding app (that's a lot of bits to keep aligned!).

Because I'm not (yet!) a proficient Vitis AI developer, I have a hard time building a complete mental of model of which bits are loaded by what using which interface and how many variations / parallel versions I can run at the time / should keep available on storage to select from :-)

My current mental model is:
- bitstream selection is more dynamic thanks to dfx-mgr, but fundamentally it's still a relatively stable thing; you typically load one large bitstreams in your target setup, perhaps it will evolve slowly over time, perhaps you have a second small one on the side; the bitstream is also very OS agnostic
- if your app was developed with Vitis-AI, you might have used some pre-built models or created your own; your app will load the models from disk on startup / through some vitis-ai libraries, and use them in combination with the corresponding bitstream

(I'm disregarding how much we're separating the pieces out in different github repositories, I'm mainly interested in where they would live on disk)

Now it kind of makes sense right now to have bitstream stuff under /lib/firmware; it's kind of OS agnostic, it's for a different architecture than the host – it will be running "elsewhere"; it's kind of stable over time too and could be involved with the boot process like device trees (albeit with dfx-mgr it's more post-loaded, so we could also consider shell.json to be a dfx-mgr payload and the rest are the accompanying files).

For models, you might to have a bank of system-wide reference models, and you might want to have custom models from the app; they would probably be tied to a specific set of vitis-ai ABIs, like the on disk file format, or the conventions to run the model; I'm guessing the model format might change between vitis-ai 2.0 and 2.5. So perhaps they would live under some kind of vitis-ai-2.0, or vitis-ai-2.5 folder for the system-wide ones, and they could be under an app specific folder as the app probably only cares to run against one vitis-ai version. Of course, it could also be named against the model ABI if that's tracked separately; like vitis-ai-model-abi-1, vitis-ai-model-abi-2 or similar.
  Perhaps /usr/share/vitis-ai-model-1/zoo-a.model, zoo-b.model etc. would make sense. /usr/share is appropriate if the model is agnostic to the architecture, but if the model depends on the architecture of the host, then it should go under /usr/lib/$triplet/vitis-ai-model-1/zoo-a.model etc.
  OS distributed apps could ship their custom/private models, or a copy of models from the zoo, under /usr/share/cool-app/my-models/app.model and their dependencies would pay attention to build/run against the vitis-ai lib that supports their model version.
  Sideloaded apps could do the same thing under /opt/what/ever/my-models/app.model. And they would want to embed a copy of vitis-ai libs under /opt/what/ever/lib/.

The OS maintenance/Vitis AI developers would then pay attention that this or that version of the runtime vitis-ai libs is compatible with this or that ABI version of the reference models and track this in dependencies. Apps would track their dependency on a specific vitis AI shared library ABI and on a specific model name from the zoo.

That still lives two problems on the top of my mind:
- tracking/enforcing model x bitstream combinations; one first step that would help here is to be able to easily get metadata about this model is for this bitstream or this bitstream contains these things and this model expect these things in the bitstream
- dynamic selection of bitstream + app(s) sets

But eh, most changes are I'm completely off base with my mental model, so I'll probably need to be educated a lot to have a sane conversation there :)

**Pull request merged**